iep-spring-leader-redis-cluster/src/test/java/com/netflix/iep/leader/redis/RedisClusterLeaderDatabaseTest.java

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright 2014-2023 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.iep.leader.redis;
+
+import com.netflix.iep.leader.api.LeaderId;
+import com.netflix.iep.leader.api.ResourceId;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.stubbing.Answer;
+import redis.clients.jedis.Connection;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.Transaction;
+import redis.clients.jedis.exceptions.JedisConnectionException;
+import redis.clients.jedis.exceptions.JedisException;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.*;
+
+public class RedisClusterLeaderDatabaseTest {
+
+  private Config config = getConfig(Collections.emptyMap());
+  private String key = "iep.leader:{MyCluster}";
+  private ResourceId resourceId = new ResourceId("MyCluster");
+  private String leader = "UT";
+
+  private RedisClusterLeaderClient cluster;
+  private RedisClusterLeaderDatabase db;
+  private Jedis client;
+  private Connection conn;
+  private Transaction transaction;
+
+  @Before
+  public void before() {
+    cluster = mock(RedisClusterLeaderClient.class);
+    JedisCluster mockCluster = mock(JedisCluster.class);
+    when(cluster.cluster()).thenReturn(mockCluster);
+    db = new RedisClusterLeaderDatabase(config, cluster);
+    client = mock(Jedis.class);
```

Comment:
Not a big fan of mocks, ideally we can keep that to a minimum and just use normal code. There is an embedded redis project, but it doesn't seem to be maintained anymore.

Reply:
Right and we'd need clustering and that gets really complicated to test properly, particularly for these failure scenarios.